File: java/src/main/java/org/htmlsanitizer/HtmlSanitizer.java
package org.htmlsanitizer;

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.MemorySegment;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;

/**
 * Cleans HTML of constructs that can lead to XSS.
 *
 * <pre>{@code
 * try (HtmlSanitizer s = new HtmlSanitizer()) {
 *     s.allowedTags().add("my-widget");
 *     String clean = s.sanitize("<div onclick=\"evil()\">hi</div>");
 * }
 * }</pre>
 *
 * <p>This class carries <b>no sanitizer logic</b>. The sanitizer core — HTML5
 * tokenizer, DOM, CSS parser, URL resolver, allow-lists — is the pure-Aether
 * {@code core/htmlsanitizer.ae}, shared by every language binding in this
 * monorepo. Everything here marshals to an {@code aether_hs_embed_*} call in
 * {@link Native}.
 *
 * <p>Instances are {@link AutoCloseable}; {@link #close()} releases the native
 * handle and the upcall stubs. Not thread-safe: the sanitizer core calls hooks
 * re-entrantly during {@code sanitize}.
 *
 * <p>Requires {@code --enable-native-access=ALL-UNNAMED} on the command line.
 */
public final class HtmlSanitizer implements AutoCloseable {

    private final Native api;
    private MemorySegment handle;

    /**
     * Upcall stubs and any other native memory tied to this sanitizer's
     * lifetime. Closed in {@link #close()}, never before: an upcall stub that
     * is freed while the sanitizer core can still call it crashes the VM.
     */
    private final Arena callbackArena = Arena.ofShared();

    // Installed handlers, kept as fields so the upcall stubs (which are
    // per-instance and dispatch back through `this`) can find them.
    private RemovingTagHandler removingTag;
    private RemovingAttributeHandler removingAttribute;
    private RemovingStyleHandler removingStyle;
    private RemovingCommentHandler removingComment;
    private PostProcessHandler postProcessNode;
    private PostProcessHandler postProcessDom;
    private FilterUrlHandler filterUrl;

    private final AllowList allowedTags;
    private final AllowList allowedAttributes;
    private final AllowList allowedCssProperties;
    private final AllowList allowedSchemes;
    private final AllowList allowedClasses;
    private final AllowList uriAttributes;

    // ---- handler interfaces ----
    //
    // For the removing* family, returning true CANCELS the removal — i.e.
    // keeps the node/attribute/property.

    @FunctionalInterface
    public interface RemovingTagHandler {
        boolean onRemovingTag(Node node, int reason);
    }

    @FunctionalInterface
    public interface RemovingAttributeHandler {
        boolean onRemovingAttribute(Node element, Attribute attribute, int reason);
    }

    @FunctionalInterface
    public interface RemovingStyleHandler {
        boolean onRemovingStyle(Node element, String name, String value, int reason);
    }

    @FunctionalInterface
    public interface RemovingCommentHandler {
        boolean onRemovingComment(Node node);
    }

    @FunctionalInterface
    public interface PostProcessHandler {
        void onPostProcess(Node node);
    }

    /** Returns the URL to use; an empty string drops the attribute. */
    @FunctionalInterface
    public interface FilterUrlHandler {
        String onFilterUrl(Node element, String rawUrl, String resolvedUrl);
    }

    // ---- lifecycle ----

    /** Load the sanitizer core and create a sanitizer with the secure defaults. */
    public HtmlSanitizer() {
        this(null);
    }

    /**
     * As {@link #HtmlSanitizer()}, but loading the sanitizer core from an explicit
     * path instead of the usual {@code $HTMLSANITIZER_LIB} / bundled /
     * loader-path search.
     */
    public HtmlSanitizer(String nativeLibPath) {
        this.api = Native.load(nativeLibPath);
        try {
            this.handle = (MemorySegment) api.hsNew.invokeExact();
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
        if (handle == null || handle.equals(MemorySegment.NULL)) {
            throw new IllegalStateException("failed to create the native sanitizer");
        }
        this.allowedTags = new AllowList(this, Native.TAGS);
        this.allowedAttributes = new AllowList(this, Native.ATTRIBUTES);
        this.allowedCssProperties = new AllowList(this, Native.CSS_PROPERTIES);
        this.allowedSchemes = new AllowList(this, Native.SCHEMES);
        this.allowedClasses = new AllowList(this, Native.CLASSES);
        this.uriAttributes = new AllowList(this, Native.URI_ATTRIBUTES);
    }

    @Override
    public void close() {
        if (handle != null) {
            try {
                // Frees the sanitizer core's callback boxes too, so no upcall stub can
                // fire after this — which is what makes closing the arena
                // immediately afterwards safe.
                api.free.invokeExact(handle);
            } catch (Throwable t) {
                throw Native.wrap(t);
            } finally {
                handle = null;
                callbackArena.close();
            }
        }
    }

    private void check() {
        if (handle == null) throw new IllegalStateException("sanitizer is closed");
    }

    Native api() {
        return api;
    }

    MemorySegment handle() {
        check();
        return handle;
    }

    /** The sanitizer core's ABI revision. */
    public int abiVersion() {
        try {
            return (int) api.abiVersion.invokeExact();
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    // ---- the main entry points ----

    /** Sanitize an HTML fragment, without resolving relative URLs. */
    public String sanitize(String html) {
        return sanitize(html, "");
    }

    /**
     * Sanitize an HTML fragment, resolving relative URLs against
     * {@code baseUrl} (which may be empty).
     */
    public String sanitize(String html, String baseUrl) {
        check();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment h = arena.allocateFrom(html == null ? "" : html);
            MemorySegment b = arena.allocateFrom(baseUrl == null ? "" : baseUrl);
            return api.takeString((MemorySegment) api.sanitize.invokeExact(handle, h, b));
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    /** Sanitize a full HTML document. */
    public String sanitizeDocument(String html) {
        return sanitizeDocument(html, "");
    }

    public String sanitizeDocument(String html, String baseUrl) {
        check();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment h = arena.allocateFrom(html == null ? "" : html);
            MemorySegment b = arena.allocateFrom(baseUrl == null ? "" : baseUrl);
            return api.takeString((MemorySegment) api.sanitizeDocument.invokeExact(handle, h, b));
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    // ---- flags ----

    /** Keep the children of a removed element instead of dropping the subtree. */
    public boolean keepChildNodes() {
        try {
            return (int) api.getKeepChildNodes.invokeExact(handle()) != 0;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public HtmlSanitizer keepChildNodes(boolean on) {
        try {
            api.setKeepChildNodes.invokeExact(handle(), on ? 1 : 0);
            return this;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    /** Allow {@code data-*} attributes through without listing each one. */
    public boolean allowDataAttributes() {
        try {
            return (int) api.getAllowDataAttributes.invokeExact(handle()) != 0;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public HtmlSanitizer allowDataAttributes(boolean on) {
        try {
            api.setAllowDataAttributes.invokeExact(handle(), on ? 1 : 0);
            return this;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    // ---- allow-lists ----

    public AllowList allowedTags() {
        return allowedTags;
    }

    public AllowList allowedAttributes() {
        return allowedAttributes;
    }

    public AllowList allowedCssProperties() {
        return allowedCssProperties;
    }

    public AllowList allowedSchemes() {
        return allowedSchemes;
    }

    public AllowList allowedClasses() {
        return allowedClasses;
    }

    public AllowList uriAttributes() {
        return uriAttributes;
    }

    // ---- callbacks ----
    //
    // Each on* builds an upcall stub bound to `this` and registers it with a
    // NULL user_data: the stub already closes over the instance, so the ABI's
    // user_data channel is spare here (other bindings use it to find their
    // handler). Passing a null handler clears the hook.

    public HtmlSanitizer onRemovingTag(RemovingTagHandler handler) {
        check();
        this.removingTag = handler;
        install(api.onRemovingTag, handler, "tagUpcall", Native.CB_REMOVING_TAG,
                MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class, int.class));
        return this;
    }

    public HtmlSanitizer onRemovingAttribute(RemovingAttributeHandler handler) {
        check();
        this.removingAttribute = handler;
        install(api.onRemovingAttribute, handler, "attributeUpcall", Native.CB_REMOVING_ATTRIBUTE,
                MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class,
                        MemorySegment.class, int.class));
        return this;
    }

    public HtmlSanitizer onRemovingStyle(RemovingStyleHandler handler) {
        check();
        this.removingStyle = handler;
        install(api.onRemovingStyle, handler, "styleUpcall", Native.CB_REMOVING_STYLE,
                MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class,
                        MemorySegment.class, MemorySegment.class, int.class));
        return this;
    }

    public HtmlSanitizer onRemovingComment(RemovingCommentHandler handler) {
        check();
        this.removingComment = handler;
        install(api.onRemovingComment, handler, "commentUpcall", Native.CB_REMOVING_COMMENT,
                MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class));
        return this;
    }

    public HtmlSanitizer onPostProcessNode(PostProcessHandler handler) {
        check();
        this.postProcessNode = handler;
        install(api.onPostProcessNode, handler, "postNodeUpcall", Native.CB_POST_PROCESS,
                MethodType.methodType(void.class, MemorySegment.class, MemorySegment.class));
        return this;
    }

    public HtmlSanitizer onPostProcessDom(PostProcessHandler handler) {
        check();
        this.postProcessDom = handler;
        install(api.onPostProcessDom, handler, "postDomUpcall", Native.CB_POST_PROCESS,
                MethodType.methodType(void.class, MemorySegment.class, MemorySegment.class));
        return this;
    }

    public HtmlSanitizer onFilterUrl(FilterUrlHandler handler) {
        check();
        this.filterUrl = handler;
        install(api.onFilterUrl, handler, "filterUrlUpcall", Native.CB_FILTER_URL,
                MethodType.methodType(MemorySegment.class, MemorySegment.class,
                        MemorySegment.class, MemorySegment.class, MemorySegment.class));
        return this;
    }

    /**
     * Register (or, for a null handler, clear) one hook.
     *
     * <p>The upcall target is bound to {@code this}, so each sanitizer gets
     * its own stub; the stub lives in {@link #callbackArena} and is released
     * only by {@link #close()}.
     */
    private void install(MethodHandle register, Object handler, String upcallName,
                         FunctionDescriptor descriptor, MethodType type) {
        try {
            if (handler == null) {
                register.invokeExact(handle, MemorySegment.NULL, MemorySegment.NULL);
                return;
            }
            MethodHandle target = MethodHandles.lookup()
                    .findVirtual(HtmlSanitizer.class, upcallName, type)
                    .bindTo(this);
            MemorySegment stub = api.linker.upcallStub(target, descriptor, callbackArena);
            register.invokeExact(handle, stub, MemorySegment.NULL);
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    // ---- upcall targets ----
    //
    // Reached only from a native trampoline. They must not declare checked
    // exceptions (the linker rejects that), so anything thrown is wrapped.
    // `ud` is the ABI's user_data — always NULL here, since the stub is bound
    // to the instance instead.

    private int tagUpcall(MemorySegment ud, MemorySegment node, int reason) {
        RemovingTagHandler h = removingTag;
        if (h == null) return 0;
        return h.onRemovingTag(new Node(api, node), reason) ? 1 : 0;
    }

    private int attributeUpcall(MemorySegment ud, MemorySegment elem,
                                MemorySegment attr, int reason) {
        RemovingAttributeHandler h = removingAttribute;
        if (h == null) return 0;
        return h.onRemovingAttribute(new Node(api, elem), new Attribute(api, attr), reason) ? 1 : 0;
    }

    private int styleUpcall(MemorySegment ud, MemorySegment elem,
                            MemorySegment name, MemorySegment value, int reason) {
        RemovingStyleHandler h = removingStyle;
        if (h == null) return 0;
        return h.onRemovingStyle(new Node(api, elem), Native.readString(name),
                Native.readString(value), reason) ? 1 : 0;
    }

    private int commentUpcall(MemorySegment ud, MemorySegment node) {
        RemovingCommentHandler h = removingComment;
        if (h == null) return 0;
        return h.onRemovingComment(new Node(api, node)) ? 1 : 0;
    }

    private void postNodeUpcall(MemorySegment ud, MemorySegment node) {
        PostProcessHandler h = postProcessNode;
        if (h != null) h.onPostProcess(new Node(api, node));
    }

    private void postDomUpcall(MemorySegment ud, MemorySegment doc) {
        PostProcessHandler h = postProcessDom;
        if (h != null) h.onPostProcess(new Node(api, doc));
    }

    private MemorySegment filterUrlUpcall(MemorySegment ud, MemorySegment elem,
                                          MemorySegment raw, MemorySegment resolved) {
        FilterUrlHandler h = filterUrl;
        // "No rewrite" — hand the resolved pointer straight back; the C
        // trampoline recognises it and neither copies nor frees it.
        if (h == null) return resolved;
        String out = h.onFilterUrl(new Node(api, elem), Native.readString(raw),
                Native.readString(resolved));
        // The sanitizer core takes ownership of this buffer and frees it with libc free.
        return api.mallocString(out);
    }
}

File: php/src/HtmlSanitizer.php
<?php

/**
 * The idiomatic PHP surface over the HtmlSanitizer core.
 *
 * Carries no sanitizer logic — every method here marshals to an
 * `aether_hs_embed_*` call in {@see Native}.
 *
 * @package HtmlSanitization
 */

declare(strict_types=1);

namespace HtmlSanitization;

use FFI;
use LogicException;
use RuntimeException;

/**
 * Cleans HTML of constructs that can lead to Cross-Site Scripting (XSS).
 *
 *     $s = new HtmlSanitizer();
 *     $s->allowedTags->add('my-widget');
 *     $clean = $s->sanitize('<div onclick="evil()">hi</div>');
 *     $s->close();
 *
 * A sanitizer is **not** safe for concurrent use — the native handle carries
 * mutable policy and hook state.
 */
final class HtmlSanitizer
{
    private FFI $ffi;

    /** @var FFI\CData|null */
    private $handle;

    /**
     * Registered FFI closures must be kept alive for as long as the sanitizer core can
     * call them. A closure that went out of scope would be freed and the
     * process would segfault on the next callback. This array is the PHP
     * equivalent of ctypes' keepalive list, and is cleared only in close(),
     * after aether_hs_embed_free has run.
     *
     * @var array<int, mixed>
     */
    private array $keepAlive = [];

    public readonly AllowList $allowedTags;
    public readonly AllowList $allowedAttributes;
    public readonly AllowList $allowedCssProperties;
    public readonly AllowList $allowedSchemes;
    public readonly AllowList $allowedClasses;
    public readonly AllowList $uriAttributes;

    /**
     * Create a sanitizer with the sanitizer core's secure defaults populated.
     *
     * @param string|null $nativeLib An explicit sanitizer core path; otherwise
     *                               $HTMLSANITIZER_LIB, then native/, then
     *                               ../core/native/, then the OS loader.
     * @throws RuntimeException when the sanitizer core cannot be loaded or created.
     */
    public function __construct(?string $nativeLib = null)
    {
        $this->ffi = Native::load($nativeLib);
        $handle = $this->ffi->aether_hs_embed_new();
        if (FFI::isNull($handle)) {
            throw new RuntimeException('htmlsanitizer: failed to create the native sanitizer');
        }
        $this->handle = $handle;

        $this->allowedTags          = new AllowList($this, Native::TAGS);
        $this->allowedAttributes    = new AllowList($this, Native::ATTRIBUTES);
        $this->allowedCssProperties = new AllowList($this, Native::CSS_PROPERTIES);
        $this->allowedSchemes       = new AllowList($this, Native::SCHEMES);
        $this->allowedClasses       = new AllowList($this, Native::CLASSES);
        $this->uriAttributes        = new AllowList($this, Native::URI_ATTRIBUTES);
    }

    /** @internal */
    public function ffi(): FFI
    {
        return $this->ffi;
    }

    /**
     * @internal
     * @return FFI\CData
     */
    public function handle()
    {
        if ($this->handle === null) {
            throw new LogicException('htmlsanitizer: sanitizer is closed');
        }
        return $this->handle;
    }

    // ---- lifecycle ----

    /** Release the native handle and every callback closure. Idempotent. */
    public function close(): void
    {
        if ($this->handle === null) {
            return;
        }
        $h = $this->handle;
        $this->handle = null;
        $this->ffi->aether_hs_embed_free($h);
        // Only now is it certain the sanitizer core can no longer invoke a hook.
        $this->keepAlive = [];
    }

    /**
     * Backstop for a dropped sanitizer; close() is the deterministic way.
     *
     * Guarded with isset(): during PHP's final shutdown GC, object destruction
     * order is not defined, so the FFI object in $this->ffi may already have
     * been torn down when this runs. Calling into it then is a segfault, not
     * an exception. A sanitizer that reaches shutdown without close() leaks
     * one native handle — which the OS reclaims on exit anyway — and that is
     * strictly better than crashing the interpreter on the way out.
     */
    public function __destruct()
    {
        if (!isset($this->ffi)) {
            return;
        }
        $this->close();
    }

    /**
     * Run $body with a fresh sanitizer, closing it afterwards even on throw.
     *
     * (Named withSanitizer rather than use: `use` is a reserved word, and
     * although PHP 7+ permits it as a method name it confuses tooling and
     * readers alike.)
     *
     * @template T
     * @param callable(HtmlSanitizer): T $body
     * @return T
     */
    public static function withSanitizer(callable $body, ?string $nativeLib = null): mixed
    {
        $s = new self($nativeLib);
        try {
            return $body($s);
        } finally {
            $s->close();
        }
    }

    // ---- the main entry point ----

    /**
     * Sanitize an HTML fragment. $baseUrl resolves relative URLs; pass ""
     * (the default) for no resolution.
     */
    public function sanitize(string $html, string $baseUrl = ''): string
    {
        return Native::takeString(
            $this->ffi,
            $this->ffi->aether_hs_embed_sanitize($this->handle(), $html, $baseUrl)
        );
    }

    /** Sanitize a full HTML document. */
    public function sanitizeDocument(string $html, string $baseUrl = ''): string
    {
        return Native::takeString(
            $this->ffi,
            $this->ffi->aether_hs_embed_sanitize_document($this->handle(), $html, $baseUrl)
        );
    }

    // ---- flags ----

    /** Keep the children of a removed element instead of dropping the subtree. */
    public function setKeepChildNodes(bool $on): self
    {
        $this->ffi->aether_hs_embed_set_keep_child_nodes($this->handle(), $on ? 1 : 0);
        return $this;
    }

    public function getKeepChildNodes(): bool
    {
        return $this->ffi->aether_hs_embed_get_keep_child_nodes($this->handle()) !== 0;
    }

    /** Allow `data-*` attributes through without listing each one. */
    public function setAllowDataAttributes(bool $on): self
    {
        $this->ffi->aether_hs_embed_set_allow_data_attributes($this->handle(), $on ? 1 : 0);
        return $this;
    }

    public function getAllowDataAttributes(): bool
    {
        return $this->ffi->aether_hs_embed_get_allow_data_attributes($this->handle()) !== 0;
    }

    /** The sanitizer core's ABI revision. */
    public function abiVersion(): int
    {
        return $this->ffi->aether_hs_embed_abi_version();
    }

    /** Where the sanitizer core .so was actually loaded from. */
    public function nativeLibraryPath(): ?string
    {
        return Native::path();
    }

    // ---- callbacks ----
    //
    // Each on* takes a PHP callable (or null to clear the hook) and returns
    // $this, so they chain. For the removing* family, returning a truthy value
    // from your handler CANCELS the removal (keeps the node/attribute/
    // property).

    /**
     * Register a trampoline with one of the ABI's seven `on_*` setters.
     *
     * PHP's FFI turns a PHP Closure into a real C function pointer when it is
     * passed to a parameter declared as a function-pointer type — which is
     * exactly how the setters are declared in {@see Native}'s cdef. Passing
     * null clears the hook.
     *
     * The Closure is stored in $keepAlive BEFORE it is registered: PHP frees
     * the generated thunk when the Closure becomes unreachable, and the sanitizer core
     * would then call into freed memory.
     *
     * user_data is unused on the PHP side — the Closure already captures the
     * handler, so there is nothing to look up. The sanitizer core's trampoline still
     * round-trips it.
     */
    private function register(string $setter, ?\Closure $trampoline): self
    {
        $h = $this->handle();
        if ($trampoline === null) {
            $this->ffi->$setter($h, null, null);
            return $this;
        }

        $this->keepAlive[] = $trampoline;
        $this->ffi->$setter($h, $trampoline, null);
        return $this;
    }

    /** `fn(Node $node, int $reason): bool` — return true to KEEP the tag. */
    public function onRemovingTag(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_removing_tag',
            $handler === null ? null : static fn ($ud, $node, int $reason): int
                => $handler(new Node($ffi, $node), $reason) ? 1 : 0
        );
    }

    /** `fn(Node $elem, Attribute $attr, int $reason): bool` — true KEEPS it. */
    public function onRemovingAttribute(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_removing_attribute',
            $handler === null ? null : static fn ($ud, $elem, $attr, int $reason): int
                => $handler(new Node($ffi, $elem), new Attribute($ffi, $attr), $reason) ? 1 : 0
        );
    }

    /** `fn(Node $elem, string $name, string $value, int $reason): bool`. */
    public function onRemovingStyle(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_removing_style',
            $handler === null ? null : static fn ($ud, $elem, $name, $value, int $reason): int
                => $handler(
                    new Node($ffi, $elem),
                    Native::borrowString($name),
                    Native::borrowString($value),
                    $reason
                ) ? 1 : 0
        );
    }

    /** `fn(Node $node): bool` — return true to KEEP the comment. */
    public function onRemovingComment(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_removing_comment',
            $handler === null ? null : static fn ($ud, $node): int
                => $handler(new Node($ffi, $node)) ? 1 : 0
        );
    }

    /** `fn(Node $node): void` — called for each node after filtering. */
    public function onPostProcessNode(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_post_process_node',
            $handler === null ? null : static function ($ud, $node) use ($handler, $ffi): void {
                $handler(new Node($ffi, $node));
            }
        );
    }

    /** `fn(Node $doc): void` — called once with the whole document. */
    public function onPostProcessDom(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_post_process_dom',
            $handler === null ? null : static function ($ud, $doc) use ($handler, $ffi): void {
                $handler(new Node($ffi, $doc));
            }
        );
    }

    /**
     * `fn(Node $elem, string $raw, string $resolved): string`
     *
     * Return the URL to use — $resolved unchanged for no rewrite, "" to drop
     * the attribute. The string is strdup'd into a buffer the sanitizer core takes
     * ownership of; you do not free it.
     */
    public function onFilterUrl(?callable $handler): self
    {
        $ffi = $this->ffi;
        return $this->register(
            'aether_hs_embed_on_filter_url',
            $handler === null ? null : static function ($ud, $elem, $raw, $resolved) use ($handler, $ffi) {
                $out = (string) $handler(
                    new Node($ffi, $elem),
                    Native::borrowString($raw),
                    Native::borrowString($resolved)
                );
                // The sanitizer core frees this, so it must come from malloc — not
                // FFI::new, whose memory PHP owns and would free a second
                // time. hs_raw_dup is the sanitizer core's own malloc'd strdup, and
                // therefore the exact counterpart of the free() that will
                // release it.
                return $ffi->hs_raw_dup($out);
            }
        );
    }

    // ---- one-shots ----

    /** Sanitize $html with the sanitizer core's defaults. */
    public static function sanitizeOnce(string $html, string $baseUrl = ''): string
    {
        return self::withSanitizer(static fn (HtmlSanitizer $s): string => $s->sanitize($html, $baseUrl));
    }

    /** Sanitize a full document with the sanitizer core's defaults. */
    public static function sanitizeDocumentOnce(string $html, string $baseUrl = ''): string
    {
        return self::withSanitizer(static fn (HtmlSanitizer $s): string => $s->sanitizeDocument($html, $baseUrl));
    }
}
